aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu

```diff
@@ -0,0 +1,506 @@
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/native/sparse/SparseUtils.h>
+#include <ATen/native/sparse/cuda/SparseCUDAApplyUtils.cuh>
+#include <ATen/native/sparse/cuda/SparseCUDABlas.cuh>
+#include <ATen/cuda/CUDAApplyUtils.cuh>
+#include <ATen/cuda/detail/IndexUtils.cuh>
+
+#include <THC/THCTensorMathPointwise.cuh>
+#include <THC/THCThrustAllocator.cuh>
+#include <THC/THCNumerics.cuh>
+#include <thrust/device_ptr.h>
+#include <thrust/sequence.h>
+#include <thrust/system/cuda/execution_policy.h>
+
+#define I_INFO(tensor) cuda::detail::getTensorInfo<int64_t, uint64_t>(tensor)
+#define V_INFO(tensor) cuda::detail::getTensorInfo<scalar_t, uint64_t>(tensor)
+
+namespace at { namespace native {
+
+// --------------------------------------------------------------------
+// Utility functions
+// --------------------------------------------------------------------
+
+#ifndef __HIP_PLATFORM_HCC__
+namespace {
+  IntTensor _to_csr_int(const LongTensor& rowIndices, int64_t dim, int64_t nnz) {
+    IntTensor csr = at::empty({dim+1}, CUDA(kInt));
+    IntTensor rowIndicesInt = at::empty({rowIndices.size(0)}, CUDA(kInt));
+    rowIndicesInt.copy_(rowIndices);
+    sparse::cuda::Xcoo2csr(rowIndicesInt.data<int32_t>(), nnz, dim, csr.data<int32_t>());
+    return csr;
+  }
+}
+#endif
+
+// NB: Deleted spaddcmul (aka addcmul_, but not actually wired up), spaddcdiv (not
+// wired at all)
+
+// --------------------------------------------------------------------
+// addmm(Tensor, SparseTensorRef, Tensor, Scalar, Scalar)  [broadcasts]
+// --------------------------------------------------------------------
+
+Tensor& s_addmm_out_sparse_dense_cuda(Tensor& r_, const Tensor& t, const SparseTensor& sparse_, const Tensor& dense, Scalar beta, Scalar alpha) {
+#ifndef __HIP_PLATFORM_HCC__
+  AT_CHECK(_check_device({sparse_, r_, t, dense}));
+  // THCudaIntTensor *csr;
+  // THCIndexTensor *indices;
+  // THCTensor *values, *r__, *dense_;
+
+  // TODO: This error message seems awfully opaque
+  AT_CHECK(sparse_._sparseDims() == 2, "matrices expected, got ", sparse_._sparseDims(), "D tensor");
+  AT_CHECK(sparse_._denseDims() == 0, "scalar values expected, got ", sparse_._denseDims(), "D values");
+  AT_CHECK(dense.dim() == 2, "matrices expected, got ", dense.dim(), "D tensor");
```

Comment:
check that dense isn't empty here.

Reply:
deferring, since empty matrices should actually multiply.